File: C++/read-n-characters-given-read4.cpp
// Forward declaration of the read4 API.
int read4(char *buf);

class Solution {
public:
    /**
     * @param buf Destination buffer
     * @param n   Maximum number of characters to read
     * @return    The number of characters read
     */
    int read(char *buf, int n) {
        int total = 0;
        bool eof = false;
        int bytes = 0;
        while(!eof && total<n) {
            bytes = read4(buf+total);
            if(bytes<4) eof = true;
            bytes = std::min(bytes, n-total);
            total += bytes;
        }
        return total;
    }
};

File: Python/read-n-characters-given-read4.py
# The read4 API is already defined for you.
# @param buf, a list of characters
# @return an integer
# def read4(buf):

class Solution:
    # @param buf, Destination buffer (a list of characters)
    # @param n,   Maximum number of characters to read (an integer)
    # @return     The number of characters read (an integer)
    def read(self, buf, n):
        
        total = 0
        bytes = 4
        while bytes==4:
            buf4 = ['']*4
            bytes = read4(buf4)
            bytes = min(bytes, n-total)
            buf[total:bytes] = buf4[:bytes]
            total += bytes
        return total
